### PR TITLE
Add boilerplate for validations

### PR DESCRIPTION
**Pull request by andreighh**
- refs https://helpdesk.eaudeweb.ro/issues/24142

### CODE REVIEW

File: core/api/serializers/cp_report.py [hunk 1]
```diff
@@ -115,36 +119,18 @@ def to_representation(self, instance):
...
-    #             raise serializers.ValidationError(
-    #                 f"Adm B, C and D are required for years before {IMPORT_DB_MAX_YEAR}"
-    #             )
-
-    #     return super().validate(attrs)
+    def validate(self, attrs):
+        if attrs.get("year") < VALIDATION_MIN_YEAR:
+            return super().validate(attrs)
+
+        validate_section_a(attrs.get("section_a"))
```
Comment (sorinstelian) [inline core/api/serializers/cp_report.py:126]: I would call a single `validate_cp_report` from `../validations/cp_reports_validations.py` and let that handle the complexity. IIRC there might be some cross-section validation, also taking into account other non-section attrs.

**Comment by andreighh** [comment]
@mereualexandra 
I tried to change the order of the operations in the `put` method (make the assignment of old country and year before the validations), but the tests fail.

Right now, I think you can skip validations if you "forge" the year, because these new validations rely on it and the initial data is passed straight to the validations. I think we should either reorder the operations or make a PUT serializer where country and year are read only. The test needs to be modified afterwards to ensure the id stays the same without a 400 response. What do you think?

**Comment by mereualexandra** [comment]
> @mereualexandra I tried to change the order of the operations in the `put` method (make the assignment of old country and year before the validations), but the tests fail.
> 
> Right now, I think you can skip validations if you "forge" the year, because these new validations rely on it and the initial data is passed straight to the validations. I think we should either reorder the operations or make a PUT serializer where country and year are read only. The test needs to be modified afterwards to ensure the id stays the same without a 400 response. What do you think?

**Comment by mereualexandra** [comment]
> @mereualexandra I tried to change the order of the operations in the `put` method (make the assignment of old country and year before the validations), but the tests fail.
> 
> Right now, I think you can skip validations if you "forge" the year, because these new validations rely on it and the initial data is passed straight to the validations. I think we should either reorder the operations or make a PUT serializer where country and year are read only. The test needs to be modified afterwards to ensure the id stays the same without a 400 response. What do you think?

Hmmm I think it's overkill to add a separate serializer for Put request just to prevent users from changing the country and year fields
We can reorder the code and maybe handle the validation for this fields in the view. In this way we can be sure that the fields will remain unchanged without adding extra complexity.
And also I think we should add a comment to be easy to understand in the future what is happeniing and why

**Comment by sorinstelian** [comment]
> @mereualexandra I tried to change the order of the operations in the `put` method (make the assignment of old country and year before the validations), but the tests fail.
> 
> Right now, I think you can skip validations if you "forge" the year, because these new validations rely on it and the initial data is passed straight to the validations. I think we should either reorder the operations or make a PUT serializer where country and year are read only. The test needs to be modified afterwards to ensure the id stays the same without a 400 response. What do you think?

my literal 2 cents: let's reorder.

**Review by raressiteavu — APPROVED** [review]
LGTM!

File: core/api/serializers/cp_report.py [hunk 2]
```diff
@@ -222,17 +202,24 @@ def create(self, validated_data):
...
-        self._create_cp_records(cp_report, validated_data.get("section_a", []), "A")
+        self._create_cp_records(cp_report, validated_data.get("section_a", []),
+                                "A")
         self._create_prices(cp_report, validated_data.get("section_c", []))
 
         if cp_report_data["year"] > IMPORT_DB_MAX_YEAR:
-            self._create_cp_records(cp_report, validated_data.get("section_b", []), "B")
-            self._create_generation(cp_report, validated_data.get("section_d", []))
-            self._create_emission(cp_report, validated_data.get("section_e", []))
+            self._create_cp_records(cp_report,
```
Comment (sorinstelian) [inline core/api/serializers/cp_report.py:210]: Even if it's not enforced in the github actions, the code looks like it's using black-style formatting.

Comment (andreighh) [reply core/api/serializers/cp_report.py:210]: Yeah, I think I forgot to set pycharm's formatting to black here, my bad.

**Review by sorinstelian — APPROVED** [review]
Good work!

**Review by mereualexandra — APPROVED** [review]
Niiiiice 😎

**Pull request merged**